dotnet/Query/Query/Query.cs

```diff
@@ -0,0 +1,558 @@
+    }
+
+    public System.TimeSpan? FindTestRunDuration()
+    {
+        // Java: if (testRunStarted == null || testRunFinished == null) return Optional.empty();
+        if (_testRunStarted == null || _testRunFinished == null)
+            return null;
+        var start = TimestampToDateTimeOffset(_testRunStarted.Timestamp);
+        var finish = TimestampToDateTimeOffset(_testRunFinished.Timestamp);
+        return finish - start;
+    }
+
+    public IList<TestStepStarted> FindTestStepsStartedBy(TestCaseStarted testCaseStarted)
+    {
+        // Java: testStepsStartedByTestCaseStartedId.getOrDefault(testCaseStarted.getId(), emptyList())
+        if (_testStepsStartedByTestCaseStartedId.TryGetValue(testCaseStarted.Id, out var steps))
+        {
+            // Return a copy for concurrency safety
+            return new List<TestStepStarted>(steps);
+        }
+        return new List<TestStepStarted>();
+    }
+
+    public IList<TestStepFinished> FindTestStepsFinishedBy(TestCaseStarted testCaseStarted)
+    {
+        // Java: testStepsFinishedByTestCaseStartedId.getOrDefault(testCaseStarted.getId(), emptyList())
+        if (_testStepsFinishedByTestCaseStartedId.TryGetValue(testCaseStarted.Id, out var steps))
+        {
+            // Return a copy for concurrency safety
+            return new List<TestStepFinished>(steps);
+        }
+        return new List<TestStepFinished>();
+    }
+
+    public IList<(TestStepFinished, TestStep)> FindTestStepFinishedAndTestStepBy(TestCaseStarted testCaseStarted)
+    {
+        // Java: findTestStepsFinishedBy(testCaseStarted).stream()
+        //   .map(testStepFinished -> findTestStepBy(testStepFinished).map(testStep -> new SimpleEntry<>(testStepFinished, testStep)))
+        //   .filter(Optional::isPresent)
+        //   .map(Optional::get)
+        //   .collect(toList());
+        var finishedSteps = FindTestStepsFinishedBy(testCaseStarted);
+        var result = new List<(TestStepFinished, TestStep)>();
+        foreach (var testStepFinished in finishedSteps)
+        {
+            var testStep = FindTestStepBy(testStepFinished);
+            if (testStep != null)
+            {
+                result.Add((testStepFinished, testStep));
+            }
+        }
+        return result;
+    }
+
+    public TestStepResult? FindMostSevereTestStepResultBy(TestCaseStarted testCaseStarted)
+    {
+        var finishedSteps = FindTestStepsFinishedBy(testCaseStarted);
+        if (finishedSteps.Count == 0)
+            return null;
+        // Find the TestStepFinished with the most severe status (highest enum value)
+        var mostSevere = finishedSteps
+            .OrderBy(f => f.TestStepResult.Status)
+            .LastOrDefault();
+        return mostSevere?.TestStepResult;
+    }
+
+    private static DateTimeOffset TimestampToDateTimeOffset(Timestamp timestamp)
+    {
+        // Java: Convertor.toInstant(timestamp)
```

Comment:
Would it make sense to put this in messages?

Reply:
For now, I've moved this into a separate extension method class.
If you'd like it moved to the Messages project along side the other Timestamp converter functions let me know.

Reply:
I think you can use the existing converter from messages, but see my other comment.